internal/core/core.go

```diff
@@ -0,0 +1,272 @@
+		}
+	}
+	return c.delta, nil
+}
+
+type validatorImpl struct {
+	mu   sync.Mutex
+	duts map[string]*checker
+}
+
+func (v *validatorImpl) check() map[string]dutCoreFiles {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	dutCores := map[string]dutCoreFiles{}
+	for _, c := range v.duts {
+		wg.Add(1)
+		go func(c *checker) {
+			defer wg.Done()
+			cores, err := c.check()
+			status := "OK"
+			if err != nil {
+				status = fmt.Sprintf("DUT %q failed to check cores: %v", c.dut.Name(), err)
+				glog.Warning(status)
+			}
+			mu.Lock()
+			dutCores[c.dut.Name()] = dutCoreFiles{
+				DUT:    c.dut.Name(),
+				Files:  cores,
+				Status: status,
+			}
+			mu.Unlock()
+		}(c)
+	}
+	wg.Wait()
+	return dutCores
+}
+
+// start starts a core file watcher for the provided DUT.
+func (v *validatorImpl) start(duts map[string]binding.DUT) map[string]dutCoreFiles {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	for k, dut := range duts {
+		glog.Infof("Registering core file checking for DUT %q", k)
+		c, err := newChecker(dut)
+		if err != nil {
+			glog.Warningf("Failed to register core file checking for DUT %q: %v", k, err)
+			continue
+		}
+		v.duts[k] = c
+	}
+	return v.check()
+}
+
+// Stop ends the validator and returns a list of all DUTs that
+// found core files.
+func (v *validatorImpl) stop() map[string]dutCoreFiles {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	return v.check()
+}
+
+func registerBefore(e *eventlis.BeforeTestsEvent) error {
+	cores := validator.start(e.Reservation.DUTs)
+	ondatra.Report().AddSuiteProperty("validator.core", "enabled")
+	report := createReport(cores)
+	ondatra.Report().AddSuiteProperty("validator.core.initial", report)
+	return nil
+}
+
+const (
+	coreFmt = `
+Delta Core Files by DUT:{{range $key, $dut := .}} 
+DUT: {{$key}}{{ range $key, $cores := $dut.Files }}
+  {{ $key }}{{ end }}{{ end }}`
+)
+
+var coreTemplate = template.Must(template.New("errorMsg").Parse(coreFmt))
+
+func createReport(d map[string]dutCoreFiles) string {
+	b := new(bytes.Buffer)
+	if err := coreTemplate.Execute(b, d); err != nil {
+		fmt.Fprintf(b, "parse error on retrieving core files: %v", err)
+	}
+	return b.String()
+}
+
+func registerAfter(e *eventlis.AfterTestsEvent) error {
+	cores := validator.stop()
+	foundCores := false
+	for _, files := range cores {
+		if len(files.Files) > 0 {
+			foundCores = true
+		}
+	}
+	report := createReport(cores)
+	if foundCores {
+		glog.Infof("Core File Check: no new cores found")
+		return fmt.Errorf("Core File check found cores:\n%s", report)
+	}
+	ondatra.Report().AddSuiteProperty("validator.core.end", report)
+	glog.Infof("Core File Check: no new cores found")
+	return nil
+}
+
+// Register will register core file watcher with the caller.
+// This will allow the event listener to fire on test module start and end.
+// All DUTs in the reservation will be monitored.
+func Register() {
+	validator = validatorImpl{
+		duts: map[string]*checker{},
+	}
+	ondatra.EventListener().AddBeforeTestsCallback(registerBefore)
```

Comment:
I'm tempted to make registerBefore and registerAfter methods on validator, so that the scope of validator can be confined to the Register function, but I also understand why those functions don't necessarily belong on validator either. I'll leave it up to you.